sdk/helper/keysutil/policy_test.go

```diff
@@ -754,3 +778,176 @@ func BenchmarkSymmetric(b *testing.B) {
+	input := []byte("the ancients say the longer the salt, the more provable the security")
+	sigAlgorithm := "pss"
+
+	tabs := make(map[int]string)
+	for i := 1; i <= 6; i++ {
+		tabs[i] = strings.Repeat("\t", i)
+	}
+
+	test_RSA_PSS := func(p *Policy, rsaKey *rsa.PrivateKey, hashType HashType, marshalingType MarshalingType) {
+		unsaltedOptions := SigningOptions{
+			HashAlgorithm: hashType,
+			Marshaling:    marshalingType,
+			SigAlgorithm:  sigAlgorithm,
+		}
+		cryptoHash := CryptoHashMap[hashType]
+		minSaltLength := p.minRSAPSSSaltLength()
+		maxSaltLength := p.maxRSAPSSSaltLength(rsaKey, cryptoHash)
+		hash := cryptoHash.New()
+		hash.Write(input)
+		input = hash.Sum(nil)
+
+		// 1. Make an "automatic" signature with the given key size and hash algorithm,
+		// but an automatically chosen salt length.
+		t.Log(tabs[3], "Make an automatic signature")
+		sig, err := p.Sign(0, nil, input, hashType, sigAlgorithm, marshalingType)
+		if err != nil {
+			t.Fatal(tabs[4], "❌ Failed to automatically sign:", err)
+		}
+
+		// 1.1 Verify this automatic signature using the *inferred* salt length.
+		autoVerify(4, t, p, input, sig, unsaltedOptions)
+
+		// 1.2. Verify this automatic signature using the *correct, given* salt length.
+		manualVerify(4, t, p, input, sig, saltOptions(unsaltedOptions, maxSaltLength))
+
+		// 1.3. Try to verify this automatic signature using *incorrect, given* salt lengths.
+		t.Log(tabs[4], "Test incorrect salt lengths")
+		incorrectSaltLengths := []int{minSaltLength, maxSaltLength - 1}
+		for _, saltLength := range incorrectSaltLengths {
+			t.Log(tabs[5], "Salt length:", saltLength)
+			saltedOptions := saltOptions(unsaltedOptions, saltLength)
+
+			verified, _ := p.VerifySignatureWithOptions(nil, input, sig.Signature, &saltedOptions)
+			if verified {
+				t.Fatal(tabs[6], "❌ Failed to invalidate", verified, "signature using incorrect salt length:", err)
+			}
+		}
+
+		// 2. Rule out boundary, invalid salt lengths.
+		t.Log(tabs[3], "Test invalid salt lengths")
+		invalidSaltLengths := []int{minSaltLength - 1, maxSaltLength + 1}
+		for _, saltLength := range invalidSaltLengths {
+			t.Log(tabs[4], "Salt length:", saltLength)
+			saltedOptions := saltOptions(unsaltedOptions, saltLength)
+
+			// 2.1. Fail to sign.
+			t.Log(tabs[5], "Try to make a manual signature")
+			_, err := p.SignWithOptions(0, nil, input, &saltedOptions)
+			if !errors.As(err, &userError) {
+				t.Fatal(tabs[6], "❌ Failed to reject invalid salt length:", err)
+			}
+
+			// 2.2. Fail to verify.
+			t.Log(tabs[5], "Try to verify an automatic signature using an invalid salt length")
+			_, err = p.VerifySignatureWithOptions(nil, input, sig.Signature, &saltedOptions)
+			if !errors.As(err, &userError) {
+				t.Fatal(tabs[6], "❌ Failed to reject invalid salt length:", err)
+			}
+		}
+
+		// 3. For each valid salt length...
+		t.Log(tabs[3], "Test all valid salt lengths")
+		for saltLength := minSaltLength; saltLength <= maxSaltLength; saltLength++ {
```

Comment:
I love how thorough this is but it's a bit brutal time wise as this portion of the test is taking around 1 and half minutes to run on my laptop :) Could we possibly test min/max and a random selection of n values perhaps of in between values?

Reply:
Yes, since any error would eventually be exposed over multiple CI runs (especially because the population of salt lengths isn't large at all), but would you have happen to have ideas for how to efficiently sample this population without replacement?

Reply:
Sadly I'm not aware of a way to easily do it within Go. Thinking a bit about this though, I'm not sure we gain much benefit of testing Go's underlying PSS implementation, so probably just validating our boundary values  (minSaltLength, maxSaltLength and say one value in between would be sufficient.

Reply:
Sounds good: should be incoming today!